File: browser-extension/icons/generate-icons.js
#!/usr/bin/env node
// Run: node icons/generate-icons.js
// Generates icon16.png, icon32.png, icon48.png, icon128.png in this directory

const zlib = require('zlib');
const fs = require('fs');
const path = require('path');

// ── CRC32 ─────────────────────────────────────────────────────────────────────
const crcTable = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = (c & 1) ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
    t[n] = c;
  }
  return t;
})();

function crc32(buf) {
  let crc = 0xFFFFFFFF;
  for (let i = 0; i < buf.length; i++) crc = (crc >>> 8) ^ crcTable[(crc ^ buf[i]) & 0xff];
  return (crc ^ 0xFFFFFFFF) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4); len.writeUInt32BE(data.length, 0);
  const t = Buffer.from(type, 'ascii');
  const c = Buffer.alloc(4); c.writeUInt32BE(crc32(Buffer.concat([t, data])), 0);
  return Buffer.concat([len, t, data, c]);
}

function encodePNG(width, height, pixels) {
  const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0); ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8; ihdr[9] = 6; // 8-bit RGBA
  const rowBytes = 1 + width * 4;
  const raw = Buffer.alloc(height * rowBytes);
  for (let y = 0; y < height; y++) {
    raw[y * rowBytes] = 0; // filter = None
    for (let x = 0; x < width; x++) {
      const s = (y * width + x) * 4;
      const d = y * rowBytes + 1 + x * 4;
      raw[d] = pixels[s]; raw[d+1] = pixels[s+1];
      raw[d+2] = pixels[s+2]; raw[d+3] = pixels[s+3];
    }
  }
  return Buffer.concat([sig, chunk('IHDR', ihdr), chunk('IDAT', zlib.deflateSync(raw, { level: 9 })), chunk('IEND', Buffer.alloc(0))]);
}

// ── Icon drawing ──────────────────────────────────────────────────────────────
// Draws a location-pin icon on an RGBA pixel buffer.
//   Background: indigo circle (#4f46e5)
//   Pin: white teardrop shape centered in the circle
//   Hole: small indigo circle (inner dot of pin)

function drawIcon(size) {
  const pixels = new Uint8Array(size * size * 4); // transparent default

  const cx = size / 2;
  const cy = size / 2;
  const bgR = size * 0.46;          // background circle radius

  const pinHeadCy = size * 0.38;    // pin head circle center Y
  const pinHeadR  = size * 0.22;    // pin head circle radius
  const holeR     = size * 0.085;   // inner hole radius
  const tipY      = size * 0.82;    // tip of pin tail

  const INDIGO = [79, 70, 229, 255];
  const WHITE  = [255, 255, 255, 255];

  function setPixel(x, y, [r, g, b, a]) {
    const i = (y * size + x) * 4;
    pixels[i] = r; pixels[i+1] = g; pixels[i+2] = b; pixels[i+3] = a;
  }

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const dx = x - cx, dy = y - cy;
      const distBg = Math.sqrt(dx * dx + dy * dy);

      // Outside background circle → transparent
      if (distBg > bgR) continue;

      // Inside background circle → indigo
      setPixel(x, y, INDIGO);

      // Pin head circle
      const dhx = x - cx, dhy = y - pinHeadCy;
      const distHead = Math.sqrt(dhx * dhx + dhy * dhy);

      if (distHead <= pinHeadR) {
        setPixel(x, y, distHead <= holeR ? INDIGO : WHITE);
        continue;
      }

      // Pin tail: teardrop below head center
      if (y > pinHeadCy && y <= tipY) {
        const t = (y - pinHeadCy) / (tipY - pinHeadCy);
        const halfW = pinHeadR * (1 - t * t);
        if (Math.abs(x - cx) <= halfW) {
          setPixel(x, y, WHITE);
        }
      }
    }
  }

  return pixels;
}

// ── Generate all sizes ────────────────────────────────────────────────────────
const sizes = [16, 32, 48, 128];
const dir = __dirname;

for (const size of sizes) {
  const pixels = drawIcon(size);
  const png = encodePNG(size, size, pixels);
  const outPath = path.join(dir, `icon${size}.png`);
  fs.writeFileSync(outPath, png);
  console.log(`✓ icon${size}.png (${png.length} bytes)`);
}

console.log('\nAll icons generated.');

File: browser-extension/icons/generate_icons.py
#!/usr/bin/env python3
"""Generate TravelPanel extension PNG icons (no external dependencies)."""
import struct, zlib, math, os

def make_png(size):
    cx = cy = size / 2.0

    def rgba(x, y):
        dx, dy = x - cx, y - cy
        dist = math.hypot(dx, dy)
        outer_r = size * 0.46
        if dist > outer_r:
            return (0, 0, 0, 0)  # transparent

        # Background: teal gradient circle
        t = dist / outer_r          # 0=centre, 1=edge
        r = int(20  + t * 8)
        g = int(184 - t * 30)
        b = int(166 - t * 25)

        # Pin body: white teardrop centred slightly above middle
        pin_cx, pin_cy = cx, cy - size * 0.06
        pin_head_r = size * 0.18

        # Inner white circle (hole in pin)
        hole_r = size * 0.07

        # Pin head circle
        head_d = math.hypot(x - pin_cx, y - pin_cy)
        if head_d <= pin_head_r:
            if head_d <= hole_r:
                return (r, g, b, 255)   # transparent hole
            return (255, 255, 255, 255)  # white pin

        # Pin tail: downward triangle below centre
        tail_top    = pin_cy + pin_head_r * 0.7
        tail_bottom = cy + size * 0.22
        if y >= tail_top and y <= tail_bottom:
            half_w = (tail_bottom - y) / (tail_bottom - tail_top) * pin_head_r * 0.55
            if abs(x - pin_cx) <= half_w:
                return (255, 255, 255, 255)

        return (r, g, b, 255)

    # Build raw bitmap
    rows = []
    for y in range(size):
        row = bytearray()
        row.append(0)   # filter: None
        for x in range(size):
            row.extend(rgba(x, y))
        rows.append(bytes(row))

    raw = b''.join(rows)

    def chunk(name, data):
        c = name + data
        return struct.pack('>I', len(data)) + c + struct.pack('>I', zlib.crc32(c) & 0xffffffff)

    ihdr = chunk(b'IHDR', struct.pack('>IIBBBBB', size, size, 8, 6, 0, 0, 0))
    idat = chunk(b'IDAT', zlib.compress(raw, 9))
    iend = chunk(b'IEND', b'')
    return b'\x89PNG\r\n\x1a\n' + ihdr + idat + iend


os.makedirs(os.path.dirname(__file__) or '.', exist_ok=True)
for sz in [16, 32, 48, 128]:
    path = os.path.join(os.path.dirname(__file__), f'icon{sz}.png')
    with open(path, 'wb') as f:
        f.write(make_png(sz))
    print(f'  ✓ icon{sz}.png')

print('Icons generated.')
